src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/ItemMapper.java

```diff
@@ -1,103 +1,35 @@
-
 package com.salesforce.dynamodbv2.mt.mappers.sharedtable.impl;
 
-import static com.amazonaws.util.CollectionUtils.isNullOrEmpty;
-
 import com.amazonaws.services.dynamodbv2.model.AttributeValue;
-import com.google.common.collect.ImmutableList;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.List;
+import com.salesforce.dynamodbv2.mt.mappers.index.DynamoSecondaryIndex;
 import java.util.Map;
+import javax.annotation.Nullable;
 
 /**
  * Maps items representing records in virtual tables so they can be read from and written to their physical table
- * counterpart according to the provided TableMapping, delegating field prefixing to the provided FieldMapper.
- * The apply() method is used to map keys in getItem, putItem, updateItem, deleteItem and for mapping item responses
- * in query and scan.
- *
- * @author msgroi
+ * counterparts.
  */
-class ItemMapper {
-
-    private final FieldMapper fieldMapper;
-    private final Map<String, List<FieldMapping>> virtualToPhysicalFieldMappings;
-    private final Map<String, List<FieldMapping>> physicalToVirtualFieldMappings;
+interface ItemMapper {
 
-    ItemMapper(FieldMapper fieldMapper, Map<String, List<FieldMapping>> virtualToPhysicalFieldMappings) {
-        this.fieldMapper = fieldMapper;
-        this.virtualToPhysicalFieldMappings = virtualToPhysicalFieldMappings;
-        this.physicalToVirtualFieldMappings = invertMapping(virtualToPhysicalFieldMappings);
-    }
-
-    /*
-     * Takes a map representing a record in a virtual table that is effectively unqualified with respect to multitenant
-     * context and returns a map representing a record in the physical table that is qualified with multitenant
-     * context appropriately.
-     *
-     * Used for adding context to GetItemRequest, PutItemRequest, UpdateItemRequest, or DeleteItemRequest objects.
+    /**
+     * Takes a virtual table record (not qualified with multitenant context) to be upserted and converts it into the
+     * corresponding physical table record (qualified with multitenant context). Used for adding context to
+     * put/update requests.
      */
-    Map<String, AttributeValue> apply(Map<String, AttributeValue> unqualifiedItem) {
-        Map<String, AttributeValue> qualifiedItem = new HashMap<>();
-        unqualifiedItem.forEach((field, attribute) -> {
-            List<FieldMapping> fieldMappings = virtualToPhysicalFieldMappings.get(field);
-            if (!isNullOrEmpty(fieldMappings)) {
-                fieldMappings.forEach(fieldMapping -> qualifiedItem.put(fieldMapping.getTarget().getName(),
-                    fieldMapping.isContextAware()
-                        ? fieldMapper.apply(fieldMapping, attribute)
-                        : attribute));
-            } else {
-                qualifiedItem.put(field, attribute);
-            }
-        });
-
-        return qualifiedItem;
-    }
+    Map<String, AttributeValue> applyForWrite(Map<String, AttributeValue> virtualItem);
```

Comment:
We used to have 2 ItemMappers in TableMapping (itemMapper and keyMapper), and I'm merging them both into this ItemMapper interface. applyForWrite is for put/update requests, when we need to convert the given virtual item into the full physical item, applying the translations for both the table primary key and any secondary indexes. See applyToKeyAttributes below for the corresponding method for what used to be keyMapper.

Reply:
I'm hoping this makes it more clear which one should be used when